models/User.go

```diff
@@ -1414,69 +1415,151 @@ func PerformMortgageTransaction(userId, stockId uint32, stockQuantity int64) (*T
+			l.Infof("PerformMortgageTransaction failed for userId = %d, stockId = %d amount = %d while retrieving total stocks", userId, stockId, stockQuantity)
 			l.Error(err)
 			return nil, err
 		}
 
-		// Sc will be negative if stocks are there in mortgage.
-		// Change sign to mean number of stocks in mortgage
-		stockCount.Sc *= -1
-
-		l.Infof("%d stocks mortgaged currently", stockCount.Sc)
+		l.Debugf("%d stocks mortgaged currently", stockCount.Sc)
 
 		if stockQuantity > stockCount.Sc {
 			l.Errorf("Insufficient stocks in mortgage. Have %d, want %d", stockCount.Sc, stockQuantity)
 			return nil, NotEnoughStocksError{}
 		}
 
 	} else {
-		MortgagePutLimitRWMutex.RLock()
-		lim := int64(MortgagePutLimit)
-		MortgagePutLimitRWMutex.RUnlock()
-		if user.Total > lim {
-			return nil, WayTooMuchCashError{}
-		}
+		l.Debugf("Mortgaging stocks in action")
+
+		db := getDB()
 
-		rate = MORTGAGE_DEPOSIT_RATE
-		l.Debugf("stockQuantity negative. Depositing stocks to mortgage @ %d%%", rate)
+		/* We don't want to include Mortgage type transaction because mortgage transactions are stored in Transactions
+		   table just for displaying/frontend purpose.*/
+
+		// stockCount = Total no. of stocks (mortgaged + unmortgaged)
+		stockCount := struct{ Sc int64 }{0}
+		sql := "Select sum(StockQuantity) as sc from Transactions where UserId=? and StockId=? and NOT Type=?"
+		err = db.Raw(sql, user.Id, stockId, MortgageTransaction.String()).Scan(&stockCount).Error
+		if err != nil {
+			l.Error(err)
+			return nil, err
+		}
 
-		stockOwned, err := getSingleStockCount(user, stockId)
+		// mortgageCount = stocksMortgaged
+		mortgageCount := struct{ Sc int64 }{0}
+		sql = "Select sum(stocksInBank) as sc from MortgageDetails where UserId=? and StockId=?"
+		err = db.Raw(sql, user.Id, stockId).Scan(&mortgageCount).Error
 		if err != nil {
 			l.Error(err)
 			return nil, err
 		}
 
-		if stockOwned < -stockQuantity {
-			l.Errorf("Insufficient stocks in ownership. Have %d, want %d", stockOwned, stockQuantity)
+		if AbsBranch(stockQuantity) > stockCount.Sc-mortgageCount.Sc {
+			l.Errorf("Insufficient stocks to mortgage. Have %d, want %d", stockCount.Sc, stockQuantity)
 			return nil, NotEnoughStocksError{}
 		}
 	}
 
-	trTotal := -int64(mortgagePrice) * stockQuantity * int64(rate) / 100
-	l.Infof("lalalalalalalala %d %d %d", trTotal, stockQuantity, int64(rate/100))
-	if int64(user.Cash)+trTotal < 0 {
-		l.Debugf("User does not have enough cash. Want %d, Have %d. Failing.", trTotal, user.Cash)
-		return nil, NotEnoughCashError{}
+	var trTotal int64
+
+	/* It is retrive action; This block will set correct trTotal and stockQuantity equal
+	to amount of stocks that could be retrieved; This block also modifies MortgageDetails
+	table, btw this table is not used anywhere except in current function */
+
+	if stockQuantity >= 0 {
+
+		db := getDB()
+		sql := "SELECT id, stocksInBank, mortgagePrice from MortgageDetails where userId=? AND stockId=? ORDER BY mortgagePrice"
```

Comment:
Don't we want to go in descending order of `mortgagePrice`? Makes more sense, since as a lender, you want to get money for higher valued assets first.

Reply:
player would want to retrieve maximum stocks out of mortgage with less cash, so retirve stocks with less mortgage price first